BUG: Ensure 'coerce' actually coerces datatypes

bashtage opened the pull request:
Changes behavior of convert objects so that passing 'coerce' will
ensure that data of the correct type is returned, even if all
values are null-types (NaN or NaT).

closes #9589

bashtage commented:
Definitely not ready, but close enough to begin discussing the changes in behavior I am proposing.

Essentially I am suggesting that if one of the flags is 'coerce' that the other must not be 'coerce' and that when 'coerce' is used then the returned array will always have the target type.

If other combinations of the keyword arguments are set to `True` (and no `'coerce'`), then the behavior is the same.  

The big change is that if `'coerce'` is passed, then no other attempts are made to convert to other types in the event that all converted values are null-types.

jreback commented:
pls rebase on master and repush had an issue with some builds

jreback commented on pandas/core/common.py:
```diff
@@ -1888,60 +1890,58 @@ def _possibly_convert_objects(values, convert_dates=True,
+        if coerce_count > 1:
+            raise ValueError("'coerce' can be used at most once.")
+
+        false_count = sum([not c for c in conversions])
```
I don't think this is worth the effort to show the warning. I would take it out (of course the ValueError above is the most important item).

jreback commented on pandas/core/common.py:
```diff
@@ -1888,60 +1890,58 @@ def _possibly_convert_objects(values, convert_dates=True,
-            new_values = _possibly_cast_to_datetime(
-                values, 'M8[ns]', coerce=True)
+    # If not object, do not convert
+    if values.dtype != np.object_:
```
use ``is_object_dtype``

jreback commented on pandas/core/common.py:
```diff
@@ -1888,60 +1890,58 @@ def _possibly_convert_objects(values, convert_dates=True,
+        return _possibly_cast_to_datetime(values, 'M8[ns]', coerce=True)
+    if convert_timedeltas == 'coerce':
+        from pandas.tseries.timedeltas import to_timedelta
+        return np.asanyarray(to_timedelta(values, coerce=True))
```
why do you need ``np.asanyarray``?

jreback commented on pandas/core/common.py:
```diff
@@ -1888,60 +1890,58 @@ def _possibly_convert_objects(values, convert_dates=True,
+        from pandas.tseries.timedeltas import to_timedelta
+        return np.asanyarray(to_timedelta(values, coerce=True))
+    if convert_numeric == 'coerce':
+        return lib.maybe_convert_numeric(values, set(), coerce_numeric=True)
```
why shouldn't these be ``elifs``? (the coerce ones)

jreback commented on pandas/core/common.py:
```diff
@@ -1888,60 +1890,58 @@ def _possibly_convert_objects(values, convert_dates=True,
+        if false_count != 2:
+            import warnings
+            warnings.warn("Soft conversion flags ignored when using 'coerce'",
+                          RedundantSettingWarning)
```
this whole part (about seeing if we have multiple 'coerces' belongs below where you actually do it. So that you have 1 place where you test if you have more than 1 coerce (ValueError), or at least 1, then branch to the coerce part, then simply return.

bashtage commented:
I had addressed all the points raised, and have two issues remaining.

1. I am using `asanyarray` since the returned type is TimedeltaIndex which is not what is expected from this function, and causes errors in other places it is called since it doesn't have all methods of an `ndarray`.  Maybe I shoudl just use `np.array`
2. The behavior in the current version is different for `convert_dates`/`convert_timestamps` and `convert_numeric`.  `c_d` and `c_ts` both will do nothing if given a mixed series -- that is, one which has some datetimes/timestamps but some non-convertibles.  `convert_numeric` will return `nan` for non-convertibles and numbers when convertible.  I am proposing to alter the behavior so that `True` will always return object-dtype when faxing mixed arrays.  Is this reasonable?

jreback commented:
1. pass ``box=False`` to ``to_timedelta`` rather than coercing (as this will simply not box it into a ``TimedeltaIndex``).

2. can you give an example (current and proposed?)

bashtage commented:
The flags are dates, numeric, timestamps.  When it is `True, False, False` or `False,False,True` then the returned series is essentially not converted since many of the objects can't be converted to datetimes or timestamps.

However, when it is `False,True,False` then the returned type is `float64` with `NaN`s for the not-convertible types.

```python
import pandas as pd
import datetime as dt
dt = dt.datetime.now()
td = dt - dt
s = pd.Series([1.0,'1.0',dt,td])

s
Out[6]: 
0                             1
1                           1.0
2    2015-06-04 10:56:04.421566
3                       0:00:00
dtype: object

s.convert_objects(True,False,False)
Out[7]: 
0                             1
1                           1.0
2    2015-06-04 10:56:04.421566
3                       0:00:00
dtype: object

s.convert_objects(False,True,False)
Out[8]: 
0     1
1     1
2   NaN
3   NaN
dtype: float64

s.convert_objects(False,False,True)
Out[9]: 
0                             1
1                           1.0
2    2015-06-04 10:56:04.421566
3                       0:00:00
dtype: object
```

Proposed correct output would not be converted -- not even the string '1.0'. **Not sure if the string should be converted or not here**

```python
s.convert_objects(False,True, False)
Out[9]: 
0                           1
1                           1.0
2    2015-06-04 10:56:04.421566
3                       0:00:00
dtype: object
```

The reason this happens in 0.16.x is that the conversion is made wiht 

```python
new_values = lib.maybe_convert_numeric(values, set()), coerce_numeric=True)
```

in all cases, so that `coerce_numeric=True` is always passed.

bashtage commented:
Some further thoughts:

The `'coerce'` options are fine -- that it `'coerce'` will always return the expected type irrespective of what is given.

The `True`/`False` options should have the following behavior:

1. For dates or timestamps, only convert if all objects are dates/timestamps
2. For numeric types, convert if *anything* is numeric-like -- either a number or a string convertible to a number.

This means that for numeric types, the only difference between `'coerce'` and `True` is when all values are `NaN` - in which case `True` will return the original array while `'coerce'` will return an array with all `NaN`s.

This still leaves the functions having a bit of *magic*, but I think being stricter would create a lot of headaches.

bashtage commented:
It should be close now.  I had to fix an issue in `groupby` so this needs some checking.  Still need to document things.

jreback commented on pandas/core/common.py:
```diff
@@ -1894,54 +1893,48 @@ def _possibly_convert_objects(values, convert_dates=True,
-        if convert_timedeltas == 'coerce':
+        # Immediate return if coerce
+        if convert_dates == 'coerce':
+            return _possibly_cast_to_datetime(values, 'M8[ns]', coerce=True)
```
you can change this to use ``to_datetime(....coerce=True)`` I think (it was that way a long time ago, but I changed it, but now for consitency prob better

jreback commented on pandas/core/groupby.py:
```diff
@@ -2939,12 +2939,13 @@ def _wrap_applied_output(self, keys, values, not_indexed_same=False):
-                    cd = 'coerce'
+                if (self._selected_obj.ndim == 2 and self._selected_obj.dtypes.isin(_DATELIKE_DTYPES).any()):
+                    result = result.convert_objects(convert_dates=False, convert_numeric=True)
+                    date_cols = [col for col, is_date in zip(result, self._selected_obj.dtypes.isin(_DATELIKE_DTYPES)) if is_date]
```
better would be

``result.select_dtypes(include=[_DATELIKE_DTYPES]).columns``

jreback commented on pandas/tests/test_groupby.py:
```diff
@@ -599,7 +599,7 @@ def f(grp):
         result = df.groupby('A').apply(f)[['C']]
         e = df.groupby('A').first()[['C']]
-        e.loc['Pony'] = np.nan
+        e.loc['Pony'] = pd.NaT
```
yeh, this was a really weird test.

jreback commented on pandas/tests/test_series.py:
```diff
@@ -5830,6 +5830,71 @@ def test_apply_dont_convert_dtype(self):
+
+        dt = datetime(2001, 1, 1, 0, 0)
+        td = dt - datetime(2000, 1, 1, 0, 0)
+        # Test coercion with mixed types
```
add a blank line

jreback commented:
@bashtage looks pretty good.

btw, I think now might be a good time to change the actual argument names as well. ( and deprecate the originals).

so instead of:

``convert_objects(self, convert_dates=True, convert_numeric=False, convert_timedeltas=True, copy=True)``

maybe?
``convert_objects(self, datetime=True, numeric=False, timedelta=True, copy=True)``

 or *maybe* we just add to directly to ``.astype(....)`` instead? (in effect make the ``dtype`` arg optional if the others are passed)

further I would not object to additional args ``floating=False,integer=False`` (to be cooperative with ``numeric``, and would clearly have to have a precedence order), unless that is MORE confusing.

bashtage commented:
I suppose my ideal API would look like

`convert_objects(datetime=True, numeric=False, timedelta=False, coerce=False, copy=True)`

so that the values for the keywords would always be `bool` and coercions would happen if `coerce` was set.

The behavior would be as follows:

1. If `coerce=True` then at most one of `datetime`, `numeric` or `timedelta` must be `True`, and the returned type will always have the correct type with all non-convertable values set to null-types.
2. If `coerce=False`, then soft conversions will be attempted where a conversion is successful if **all** values can be converted to a particular type.  If any values are not-convertible, then the original object array is returned.

The current version satisfies 1 but not 2 since the behavior of `convert_numeric=True` is to always convert if at least one value is convertible to numeric, while `convert_dates` and `convert_timedelta` will only convert if all are convertible.

I'm also not so sure about the `copy` flag since any conversion will be changing the underlying blocks since a columns will be changing from object to another type.  What would `copy=False` be expected to do?  Replace "inplace"?  Return the same object with different blocks?

jreback commented:
``copy=True`` is the default and in reality should always be ``True``. Only in very very specific circumstances (and internally only) do I support ``copy=False``. Its not really 'public' in that sense, so maybe should be ``_copy=True`` to mark it as an internal option (or could just dispense with flag, and make another func for internal consumption).

bashtage commented:
LMK what you think of this version of the API.  It is not

```python
df.convert_objects(datetime=True, numeric=False, timedelta=True, coerce=False, copy=True)
```

and if `coerce=True`, then only one of the three options can be `True`.  Right now using `coerce` is a little tedious since it require setting one of `datetime` or `timedelta` to be `False`.  Have though about making `timedelta=False` as a default.

bashtage commented:
@jreback It is green now so the major question is whether the API change is reasonable or not?  Will hold off on doc changes until then.

jreback commented on pandas/core/common.py:
```diff
@@ -1883,65 +1883,55 @@ def _maybe_box_datetimelike(value):
-                              convert_timedeltas=True):
+def _possibly_convert_objects(values, datetime=True,
+                              numeric=True,
+                              timedeltas=True,
```
since changing, I think this should be ``timedelta`` no? (so all singular)

jreback commented on pandas/core/generic.py:
```diff
@@ -2423,22 +2423,23 @@ def copy(self, deep=True):
-        convert_numeric : boolean, default False
+        datetime : boolean, default True
+            If True, convert to date where possible.
+        numeric : boolean, default False
```
need to use ``deprecate`` decorator (on the old kw's). These will show a warning and then get reassigned to the new kws. So it will still work.

jreback commented:
@bashtage I am ok with the API change (small typo on the ``timedeltas`` should be ``timedelta``). This will have to be in 0.17.0 as we will deprecate the old. (0.16.2 coming later this week).

bashtage commented:
@jreback Should the `@deprecated` be in a separate PR so it can be merged before this one?

jreback commented:
@bashtage no, in this PR. It doesn't make sense to deprecate something w/o replacing it. (unless its complete vestige), in this case you are renaming things, the canonical reason to use deprecate

bashtage commented:
@jreback I think this is ready when you are.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -39,6 +39,35 @@ Backwards incompatible API changes
+=====================   =============
+``convert_dates``       ``datetime``
+``convert_numeric``     ``numeric``
+``convert_timestamp``   ``timestamp``
```
this should be timedeltas

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -39,6 +39,35 @@ Backwards incompatible API changes
+
+  .. code-block:: python
+
+    In [2]: df = pd.DataFrame({'i': ['1','2'], 'f': ['apple', '4.2']})
```
so put the frame creation above in a separate ipython-block

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -39,6 +39,35 @@ Backwards incompatible API changes
+
+  .. code-block:: python
+
+    In [2]: df = pd.DataFrame({'i': ['1','2'], 'f': ['apple', '4.2']})
```
show the results of this code (no need to show the original); you may also way to show
``df.convert_objects(dates=False, numeric=True, coerce=True).dtypes``

jreback commented on pandas/core/common.py:
```diff
@@ -1883,65 +1883,55 @@ def _maybe_box_datetimelike(value):
-def _possibly_convert_objects(values, convert_dates=True,
-                              convert_numeric=True,
-                              convert_timedeltas=True):
+def _possibly_convert_objects(values, datetime=True,
```
separate line for the 2nd argument too

jreback commented on pandas/core/common.py:
```diff
@@ -1883,65 +1883,55 @@ def _maybe_box_datetimelike(value):
+                                                   set(),
+                                                   coerce_numeric=True)
+            # If all NaNs, then do not-alter
+            values = converted if not isnull(converted).all() else values
```
*maybe* want to move this into ``maybe_convert_numeric`` directly

bashtage replied:
The issue here is that if it isn't numeric and coerce=False, it will raise and do nothing, but if it is not numeric but is a timestamp then it will get nan-filled.  This is why this logic is here.

jreback commented on pandas/core/frame.py:
```diff
@@ -3955,7 +3957,8 @@ def append(self, other, ignore_index=False, verify_integrity=False):
             other = other.reindex(combined_columns, copy=False)
             other = DataFrame(other.values.reshape((1, len(other))),
-                              index=index, columns=combined_columns).convert_objects()
+                              index=index, columns=combined_columns).convert_objects(datetime=True,
```
put a new line before the ``columns=...).convert_object``

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -39,6 +39,35 @@ Backwards incompatible API changes
 Other API Changes
 ^^^^^^^^^^^^^^^^^
 
+- ``DataFrame.convert_objects`` keyword arguments have been shortened:
```
needs the issue number that is being closed

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2427,22 +2427,26 @@ def copy(self, deep=True):
-            conversion, with unconvertible values becoming NaT.
+        timedelta : boolean, default True
+            If True, convert to timedelta where possible.
+        coerce : booleam, default False
```
booleam -> boolean

jorisvandenbossche commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -45,6 +45,45 @@ Other API Changes
+keyword argument to ``'coerce'`` instead of ``True``, as in ``convert_dates='coerce'``.
+
+  Previously data was coerced to a common dtype before serialisation, which for
+  example resulted in integers being serialised to floats:
```
I don't understand this: when I run the example below with 0.16.2, I get integers:

```
In [19]: pd.__version__
Out[19]: '0.16.2'

In [20]: df = pd.DataFrame({'i': ['1','2'], 'f': ['apple', '4.2']})

In [21]: df.convert_objects(convert_numeric='coerce')
Out[21]:
     f  i
0  NaN  1
1  4.2  2

In [22]: df.convert_objects(convert_numeric='coerce').dtypes
Out[22]:
f    float64
i      int64
dtype: object
```

jorisvandenbossche commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -45,6 +45,45 @@ Other API Changes
 - Enable writing Excel files in :ref:`memory <_io.excel_writing_buffer>` using StringIO/BytesIO (:issue:`7074`)
 - Enable serialization of lists and dicts to strings in ExcelWriter (:issue:`8188`)
 
+- ``DataFrame.convert_objects`` keyword arguments have been shortened. (:issue:`10265`)
```
Can you put everything you added into a separate section above 'Other API Changes' (with the same header level as that one), like "Changes to convert_objects" ?

jorisvandenbossche commented on pandas/core/common.py:
```diff
@@ -1883,65 +1883,56 @@ def _maybe_box_datetimelike(value):
+    if coerce:
+        coerce_count = sum((datetime, numeric, timedelta))
+        if coerce_count > 1:
+            raise ValueError("Only one of datetime, numeric or timedelta "
```
Can you put the keyword arguments in single quotes? (that makes it a bit clearer it are arguments), like ``"ONly one of 'datetime', ...``

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -1550,9 +1550,10 @@ but occasionally has non-dates intermixed and you want to represent as missing.
                  '20010105'], dtype='O')
    s
-   s.convert_objects(convert_dates='coerce')
+   s.convert_objects(datetime=True,coerce=True)
```
space after the comma for pep8

jorisvandenbossche commented:
Some other questions:

1. Now, the `convert_numeric` set to True or 'coerce' does not seem to differ (it always coerces?)

  ```
In [30]: s = Series(['a', 'b', '1'])

In [31]: s.convert_objects(convert_numeric=True)
Out[31]:
0   NaN
1   NaN
2     1
dtype: float64

In [32]: s.convert_objects(convert_numeric='coerce')
Out[32]:
0   NaN
1   NaN
2     1
dtype: float64
```
  Is this still the same? (I assume so from the tests) If so, this should maybe be more clear from the docs?

2. The fact that you now have to do `df.convert_objects(datetime=False, numeric=True, coerce=True)` (as only one of the datetime, numeric or timedelta can be True when coerce=True). Is there nothing we can do about this? (because it is rather ugly?) 
  Maybe we could set `datetime=False` as the default? (we can detect this case if all three are False, and then use `datetime=True` (and possibly raise a warning))

3. Currenly, when having no convertible values, that original is always returned, also when specifying 'coerce':

  ```
In [26]: s = Series(['a', 'b', 'c'])

In [27]: s.convert_objects(convert_numeric=True)
Out[27]:
0    a
1    b
2    c
dtype: object

In [28]: s.convert_objects(convert_numeric='coerce')
Out[28]:
0    a
1    b
2    c
dtype: object
  ```

  With this PR, from the tests it seems that this now gives all NaNs? I think this is a reasonable change, but this should then also be listed in the API changes in whatsnew?

jreback commented:
@jorisvandenbossche ``.convert_objects`` generally is not used very much as almost all coercion is already done automatically, or via explicit converters like ``to_datetime/to_timedelta``. Aside from some internal use, the main use is the ``to_numeric`` which is basically an ``.astype`` with that handles non-convertibles.

I suppose another option is to make ``.convert_objects`` private, and add ``pd.to_numeric`` to keep things clear; a third option is to add a ``coerce`` argument to ``.astype`` to allow have a nicer API.

Lastly, I suppose there is a need for a function to just 'take this object series and type it correctly'. Hence having all of these in one function. I don't think this actually a common case, but at times I suppose could be useful.

jorisvandenbossche commented:
@jreback Although maybe not that much, but it *is* used (for example, StackOverflow gives quite some results if you search for that) 
And for the case you have (for whatever reason) mixed numeric and strings, this is a useful function no? (more flexible as `astype`)

I agree that eg adding `coerce` to `astype` would be a useful enhancement, but nonetheless, we should take care at `convert_objects` as well. And even it is regarded as internal, it should be clear what it does, and so my questions above should be answered :-)

jorisvandenbossche commented:
@bashtage BTW, you have a failing test because "3.1415" seems to be parsed as a timedelta. With master:

```
In [5]: pd.to_timedelta('3.1415')
Out[5]: Timedelta('0 days 00:00:00.141500')
```

which seems wrong in any case (@jreback maybe related to the changes there #10396 ?)

jreback commented:
@jorisvandenbossche you are looking at old posts on SO, this WAS used quite a but should not be necessary in quite a few versions, except in a small number of cases.

Since we have a proposed API change (to make a better naming convention), though which is back-compat. I think it might be nicer to simply move this to ``.astype`` as that is the use case. 

would be also ok with renaming with to ``.convert``

jreback commented:
``Timedelta('3.1415')`` is wrong in 0.16.2 and master. So will open a separate issue for this. Was ok with accepting single integers e.g. ``Timedelta('3')`` which is 3ns. But now that I think about it these should raise (I don't reallly remember why I originally did that). They don't have units in this case (different story when you are explicity passing to ``to_timedelta``). So its ambiguous.

bashtage commented:
@jorisvandenbossche This failure is my reward for religously rebasing :-/

bashtage commented:
## API Design

### The reason for `convert_objects`
Inside pandas, convert objects is frequently used to re-convert columns back to datetime after some aggregation operation, for example a `grouby`.  

This is why the proposed API is so ugly: I wanted to minimize internal refactoring of `convert_objects` and having defaults `datetime=True` with everything else `False` achieved this goal.

Externally the main reason to use `convert_objects` **without** `coerce=True` is to magically convert a DataFrame of string values to one of the three supported formats.  The logic is as follows:

```
for col in data:
    if all of data[col] is convertable to datetime:
        convert data[col]) to datetime
    elif any of data[col] is convertable to numeric
        convert data[col] to numeric (int, float as appropriate)
   elif all of data[col] is convertible to timedelta:
       convert data([col]) to timedelta
````

This is strange since the behavior of partially convertable columns is different depending on the type.  This is how is is in 0.16 and I didn't want to alter this behavior since this magic is used in various places.

### What should the API look like
For single type conversions, `Series` and `DataFrame` should use `astype(*,coerce=True)` which would do exactly as told - no magic, no guessing, just 100% consistent.

This would mean `convert_objects` could have the `coerce` keyword **removed** and so would only be for soft conversions.

bashtage commented:
> Now, the convert_numeric set to True or 'coerce' does not seem to differ (it always coerces?)

I think this is correct - if the column is all strings not parsable to numbers, this will return the original data irrespective of `coerce`.  More-or-less the logic of `convert_objects` in 0.16 is deeply flawed (there are some errors in the code).  For example

https://github.com/pydata/pandas/pull/10265/files#diff-21cb6655a9828d526610249e3d6cb16eL1921

cannot do anything since new_values isn't even set.  

Coverage of this internal function is not great so no wonder this slipped through.

> The fact that you now have to do df.convert_objects(datetime=False, numeric=True, coerce=True) (as only one of the datetime, numeric or timedelta can be True when coerce=True). Is there nothing we can do about this? (because it is rather ugly?)
Maybe we could set datetime=False as the default? (we can detect this case if all three are False, and then use datetime=True (and possibly raise a warning))

I am hesitant to do this since there are lots of internal places it is assumed `convert_dates=True` in 0.16.

jreback commented:
ok, then let's deprecate ``raise_on_error`` in ``.astype`` and replace it with ``coerce`` (you may need to slightly adjust the logic in ``core/internals/_astype``).

Then ``convert_objects`` is much simpler (only does soft conversions).

bashtage commented:
Ok, will work on this in the next few days

jreback commented:
>The fact that you now have to do df.convert_objects(datetime=False, numeric=True, coerce=True) (as only one of the datetime, numeric or timedelta can be True when coerce=True). Is there nothing we can do about this? (because it is rather ugly?)
Maybe we could set datetime=False as the default? (we can detect this case if all three are False, and then use datetime=True (and possibly raise a warning))
>>I am hesitant to do this since there are lots of internal places it is assumed convert_dates=True in 0.16.

This is simply not possible, nor desired. ``.convert_objects`` is NOT meant for public consumption except in a very special case (naming safe-astyping). If this is moved to ``.astype`` then no reason to have this public anymore.

bashtage commented:
Having started looking at `astype` I'm less sure about what I shoudl expect to happen.

For example, what should happen if I have 
```
df = pd.DataFrame({'a':[1.3,3.1415]})
df.astype(np.int, coerce=True)
```
Converting object dtype to other things makes sense, but `coerce` doesn't seem to be generally applicable without lots of arbitrary choices (like how to change floats to integers).

I suppose the big advantage of `convert_objects` is that it is only for object columns, and not for anything else.  

Maybe `astype(*,coerce_object=True)` as a special case, or just `coerce` but with a note/warning that `coerce` is only applicable when when the input `dtype` is `object`?

bashtage commented:
My previous example isn't useful.  This is maybe clearer:

```
df = pd.DataFrame({'a':['2015-01-01','2015-01-02']}).astype(np.datetime64)
df.astype(np.float64, coerce=True)
```
or similarly

```
df = pd.DataFrame({'a':[1.0,np.nan]})
 df.astype(np.int64, coerce=True)
```

What would be expected?  It seems a general dtype to dtype conversion with coercion might be very difficult and complex.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -40,11 +40,54 @@ Backwards incompatible API changes
+    0  apple  1
+    1    4.2  2
+
+    In [3]: df.convert_objects(datetime=False, numeric=True, coerce=True)
```
prob more clear to leave out the ``datetime`` arg (as its ``False`` by default anyhow)

jreback commented on pandas/core/generic.py:
```diff
@@ -2427,22 +2427,26 @@ def copy(self, deep=True):
+    @deprecate_kwarg(old_arg_name='convert_dates', new_arg_name='datetime')
+    @deprecate_kwarg(old_arg_name='convert_numeric', new_arg_name='numeric')
+    @deprecate_kwarg(old_arg_name='convert_timedeltas', new_arg_name='timedelta')
+    def convert_objects(self, datetime=True, numeric=False,
```
hmm, I think to be consistent, let's make ``datetime=False`` by-default (you can simply specify it as ``True`` where needed for soft-conversions), but from a user perspective this would be confusing. Then I would raise if all three are ``False`` (IOW the user didn't specify any type of conversions)

jreback commented on pandas/core/generic.py:
```diff
@@ -2427,22 +2427,26 @@ def copy(self, deep=True):
-            If True, convert to timedelta where possible. If 'coerce', force
-            conversion, with unconvertible values becoming NaT.
+        timedelta : boolean, default True
+            If True, convert to timedelta where possible.
```
same here, let's default to ``False``

jreback commented on pandas/tests/test_series.py:
```diff
@@ -5875,38 +5875,103 @@ def test_apply_dont_convert_dtype(self):
 
     def test_convert_objects(self):
+        # Tests: All to nans, coerce, true
+        # Test coercion returns correct type
```
add the issue number as a comment

bashtage commented:
I am giving up on `astype` as a general solution for now.  

I have also tried setting all defaults to `False` so that you can use things like 

```
df.convert_objects(datetime=True)
df.convert_objects(numeric=True)
df.convert_objects(timedelta=True)
```

Assuming you think all False is reasonable, I can update the docs to reflect this.

jreback commented:
yep that is ok

I suppose all false is an error but it's kind of innocuous so prob safe to ignore (but put in an explicit test - will just return the input)

jreback commented on pandas/tslib.pyx:
```diff
@@ -2416,6 +2416,8 @@ cdef inline parse_timedelta_string(object ts, coerce=False):
 
         if (len(number) or len(frac)) and not len(unit) and current_unit is None:
+            if coerce:
+                return iNaT
```
hmm, what exactly is testing this?

as this is not allowed (and ``Timedelta`` doesn't have a coerce option itself)
```
In [1]: pd.Timedelta('1.')
ValueError: no units specified
```

bashtage replied:
`coerce` should `coerce` - this was raising when trying to convert `3.1415` when `coerce=True` and was incorrectly raising a ValueError.

The recent commit to tslib broke tests in this PR.

bashtage replied:
See https://github.com/pydata/pandas/pull/10265/files#diff-933180870dd152b775952fc54a99dba7R5928

for the test that was failing.

jreback replied:
ahh ok

jreback commented on doc/source/basics.rst:
```diff
@@ -1522,23 +1522,25 @@ then the more *general* one will be used as the result of the operation.
-pass ``convert_numeric=True``. This will force strings and numbers alike to be numbers if possible, otherwise
-they will be set to ``np.nan``.
+:meth:`~DataFrame.convert_objects` is a method to try to force conversion of
+types from the ``object`` dtype to other types. To try converting specific
```
Can you add a ``::note`` here that the API changed in 0.17.0

jreback commented on doc/source/basics.rst:
```diff
@@ -1550,9 +1552,10 @@ but occasionally has non-dates intermixed and you want to represent as missing.
-In addition, :meth:`~DataFrame.convert_objects` will attempt the *soft* conversion of any *object* dtypes, meaning that if all
+Without passing ``coerce=True``, :meth:`~DataFrame.convert_objects` will attempt
+the *soft* conversion of any *object* dtypes, meaning that if all
 the objects in a Series are of the same type, the Series will have that dtype.
```
add something like: but will not coerce strings to dates for example

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -46,12 +46,58 @@ Backwards incompatible API changes
 
+Changes to convert_objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+- ``DataFrame.convert_objects`` keyword arguments have been shortened. (:issue:`10265`)
```
I think you have an extra space here

bashtage replied:
I'm not sure I see it.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -46,12 +46,58 @@ Backwards incompatible API changes
+keyword argument ``coerce=True``.  Previously types were coerced by setting a
+keyword argument to ``'coerce'`` instead of ``True``, as in ``convert_dates='coerce'``.
+
+  Previously data was coerced to a common dtype before serialisation, which for
```
this is a bit of a non-sequitor. This doesn't have anything to do with serialization. Or are you referring to something else here?

bashtage replied:
I'm not sure what that was doing there.  C&P error I think.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -46,12 +46,58 @@ Backwards incompatible API changes
+
+  Now the ``coerce`` keyword must be explicitly used.
+
+  .. code-block:: python
```
make this an ipython-block so the code runs

I would put the initial frame at the top in another block (so its not repeated in the code-block), e.g. something like

```
.. ipython:: python

    df = pd.DateFrame(.....)
    df
```

bashtage replied:
Presumably the old code which doesn't work any more should just be in a standard python block?

```
  .. code-block:: python
  
    In [2]: df.convert_objects(convert_numeric='coerce')
```

jreback commented on pandas/core/common.py:
```diff
@@ -1887,65 +1887,62 @@ def _maybe_box_datetimelike(value):
-            # soft-conversion
-            values = lib.maybe_convert_objects(values)
+    elif not is_object_dtype(values.dtype):
+        # If not object, do not attempt conversion
```
need to handle the ``copy`` arg here. All of the other conversions copy I think. (need a test for this as well)

bashtage replied:
This function doesn't have a copy keyword - this is handled by the caller.

```
_possibly_convert_objects(values,datetime=True,numeric=True,
                              timedelta=True,coerce=False)
```

jreback replied:
this is a bug. This needs handling either IN ``_possibly_convert_objects`` or, handle [here](https://github.com/pydata/pandas/blob/master/pandas/core/internals.py#L1458). IOW, fall thru on the if/else, then copy if indicated. Though if some conversion is done then it is a new object. So this needs to be copied is if you don't pass ANY options (and the copy flag is True) OR it is all nan after conversion (and so you don't actually convert anything), e.g. L1944 (below)

bashtage replied:
I think the bug is in `convert` -- it is the last function to see `copy=?` but doesn't do anything with it.

jreback replied:
yep could be

bashtage replied:
Pushed copy one level deeper so I can trap these.

jreback commented on pandas/core/common.py:
```diff
@@ -1887,65 +1887,62 @@ def _maybe_box_datetimelike(value):
+    conversion_count = sum((datetime, numeric, timedelta))
+    if conversion_count == 0:
+        import warnings
+        warnings.warn('Must explicitly pass type for conversion. Original '
```
I think I said to add this, but rethinking a bit. @bashtage what's your opinion on this?

bashtage replied:
I think it should stay in until around 0.19.  For example, if you have a script that uses 

```
df.convert_objects()
```

in 0.16, you will get soft conversion to datetime.  If you run the same script in 0.17, you will get nothing, so it seems a warning is reasonable [Even if this code is not heavily used]

jorisvandenbossche replied:
Indeed, probably better for now to keep the same behaviour (easily detectable if all are False, act like datetime=True and raise a warning?)

bashtage commented:
@jreback I think I have hit all of your points, so hopefully this is ready.

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -1522,23 +1522,29 @@ then the more *general* one will be used as the result of the operation.
-they will be set to ``np.nan``.
+.. note::
+
+The syntax of :meth:`~DataFrame.convert_objects`  changed in 0.17.0.
```
This line has to be intended to belong to the note.

Further, can you add a link to the whatsnew notes to have an explanation what changed

bashtage replied:
Have tried to clarify

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -1550,10 +1556,13 @@ but occasionally has non-dates intermixed and you want to represent as missing.
+the *soft* conversion of any *object* dtypes, meaning that if all
 the objects in a Series are of the same type, the Series will have that dtype.
+Setting ``coerce=True`` will not *convert* - for example, a series of string
+dates will not be converted to a series of datetimes.
```
This sentence is not fully clear to me. What do you want to say exactly?

jorisvandenbossche commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -46,12 +46,51 @@ Backwards incompatible API changes
 .. _whatsnew_0170.api_breaking.other:
 
+Changes to convert_objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
you can add a link name, eg ``.. _whatsnew_0170.api_breaking.convert_objects:`` to refer to

jorisvandenbossche commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -46,12 +46,51 @@ Backwards incompatible API changes
+    df.convert_objects(numeric=True, coerce=True)
+
+- The new default behavior for ``DataFrame.convert_objects`` is to do nothing,
+and so it is necessary to pass at least one conversion target when calling.
```
maybe add that previously the default was to do soft conversion of dates? So that the previous `df.convert_objects()` has to be replaces by `df.convert_objects(datetime=True)`

jorisvandenbossche commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -46,12 +46,51 @@ Backwards incompatible API changes
+
+  .. ipython:: python
+
+    df.convert_objects(numeric=True, coerce=True)
```
This example seems a bit strange, as for `numeric`, the setting of `coerce` does not matter?

bashtage replied:
Updated to include case where coerce matters

jreback replied:
maybe add an example of what a usage with coerce did when u have 0 and 1 of the values being converted  (eg no convert / convert ) - this is the example that could break code so want to make it prominent

bashtage replied:
Added one.

Commit:
CLN: PEP 8 improvements

bashtage commented:
@jorisvandenbossche I think I got all your comments.

jreback commented on doc/source/basics.rst:
```diff
@@ -1522,23 +1522,31 @@ then the more *general* one will be used as the result of the operation.
+:meth:`~DataFrame.convert_objects` is a method to try to force conversion of
+types from the ``object`` dtype to other types. To try converting specific
+types that are *number like*, e.g. could be a string that represents a number,
+pass ``numeric=True``. The force the conversion, add the keword argument
```
To force the conversion

jreback commented on doc/source/basics.rst:
```diff
@@ -1550,10 +1558,15 @@ but occasionally has non-dates intermixed and you want to represent as missing.
+*soft* conversion of any *object* dtypes, meaning that if all
 the objects in a Series are of the same type, the Series will have that dtype.
+Note that setting ``coerce=True`` does not *convert* arbitrary types to either
+``datetime64[ns]`` or ``timedelta[ns]``. For example, a series containing string
```
timedelta64[ns]

jreback commented on pandas/core/common.py:
```diff
@@ -1887,65 +1887,68 @@ def _maybe_box_datetimelike(value):
+    if conversion_count == 0:
+        import warnings
+        warnings.warn('Must explicitly pass type for conversion. Defaulting to '
+                      'pre-0.17 behavior where datetime=True, numeric=True, '
```
nice. I added this to the deprecation issue so can remove this at some point.

jreback commented on pandas/io/wb.py:
```diff
@@ -155,7 +155,7 @@ def download(country=['MX', 'CA', 'US'], indicator=['NY.GDP.MKTP.CD', 'NY.GNS.IC
         out = out.drop('iso_code', axis=1)
         out = out.set_index(['country', 'year'])
-        out = out.convert_objects(convert_numeric=True)
+        out = out.convert_objects(datetime=True, numeric=True)
```
FYI, I think ``pandas-datareader`` needs this change (conditionally on the pandas version)

jreback replied:
side note, can you post an issue on datareader (and link back to here). thxs.

jreback commented on pandas/tests/test_series.py:
```diff
@@ -5912,39 +5913,105 @@ def test_apply_dont_convert_dtype(self):
+
+        # test pass-through and non-conversion when other types selected
+        s = Series(['1.0','2.0','3.0'])
+        results = s.convert_objects(True,True,True)
```
can you put the kw args to make this explicit what you are doing (in several tests)

jreback commented:
@bashtage couple of minor doc corrections. I think ready to go. pls make those fixes and ping when green.

bashtage committed:
BUG: Ensure 'coerce' actually coerces datatypes

Changes behavior of convert objects so that passing 'coerce' will
ensure that data of the correct type is returned, even if all
values are null-types (NaN or NaT).

closes #9589

bashtage commented:
@jreback ready

jreback commented:
looks good! ping on green!

bashtage commented:
@jreback Green.

jreback commented:
ty @bashtage !

putting in a big PR often requires several iterations, so thanks for the patience!

jreback commented:
pls review the built docs (usually take a few hours for travis to do it), and issue any corrections you see/need.
